### PR TITLE
Fix the issue where we load segment metadata directly from metadata file

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Add a API to directly load segment metadata from metadata file

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/SegmentMetadataImpl.java [hunk 1]
```diff
@@ -257,6 +257,33 @@ public SegmentMetadataImpl(RealtimeSegmentZKMetadata segmentMetadata, Schema sch
     setSchema(schema);
   }
 
+  private SegmentMetadataImpl() {
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/SegmentMetadataImpl.java]: not required if you have other constructors

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/SegmentMetadataImpl.java [hunk 2]
```diff
@@ -257,6 +257,33 @@ public SegmentMetadataImpl(RealtimeSegmentZKMetadata segmentMetadata, Schema sch
...
+  }
+
+  /**
+   * Load segment metadata directly from metadata file.
+   */
+  public static SegmentMetadataImpl fromMetadataFile(@Nonnull File metadataFile)
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/SegmentMetadataImpl.java]: Don't mix constructors with static factory method.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 3]
```diff
@@ -629,7 +629,8 @@ protected SegmentMetadataImpl extractSegmentMetadata(final String rawTableName,
           .unTarOneFile(new FileInputStream(segFile), V1Constants.MetadataKeys.METADATA_FILE_NAME);
       metadataPath = FileSystems.getDefault().getPath(baseDir, segmentNameStr + ".metadata");
       Files.copy(is, metadataPath);
-      segmentMetadata = new SegmentMetadataImpl(new File(metadataPath.toString()));
+      // TODO: crc is not loaded because it's inside creation.metadata file
+      segmentMetadata = SegmentMetadataImpl.fromMetadataFile(new File(metadataPath.toString()));
```
Comment (antumbde) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: why not untar and load segment metadata. This is a test so we can

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 4]
```diff
@@ -629,7 +629,8 @@ protected SegmentMetadataImpl extractSegmentMetadata(final String rawTableName,
           .unTarOneFile(new FileInputStream(segFile), V1Constants.MetadataKeys.METADATA_FILE_NAME);
       metadataPath = FileSystems.getDefault().getPath(baseDir, segmentNameStr + ".metadata");
       Files.copy(is, metadataPath);
-      segmentMetadata = new SegmentMetadataImpl(new File(metadataPath.toString()));
+      // TODO: crc is not loaded because it's inside creation.metadata file
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: If the crc is not set right, then will the servers reload all segments all the time? This function worked right before. What happened? Why can't we fix it (even sub-optimally) while getting the crc correctly?

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java (hunk 3 above)
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: This is not a test... I will remove the new static factory method and uniform the behavior (always load from a directory)

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 5]
```diff
@@ -610,34 +613,50 @@ public static int getRealtimeTableFlushSize(AbstractTableConfig tableConfig) {
...
+      // Extract creation.meta
+      InputStream creationMetaInputStream =
+          TarGzCompressionUtils.unTarOneFile(new FileInputStream(new File(segFileStr)),
+              V1Constants.SEGMENT_CREATION_META);
+      Preconditions.checkNotNull(creationMetaInputStream, "%s does not exist", V1Constants.SEGMENT_CREATION_META);
+      Path creationMetaPath = FileSystems.getDefault().getPath(tempMetadataDirStr, V1Constants.SEGMENT_CREATION_META);
+      Files.copy(creationMetaInputStream, creationMetaPath);
+
+      // Load segment metadata
+      return new SegmentMetadataImpl(new File(tempMetadataDirStr));
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: reuse tempMetadataDir

**Pull request merged**